### DISCUSSION TITLE
Types for `get()` method of a reference?

### DISCUSSION THREAD

**Discussion by davidatsurge** [mobxjs/mobx-state-tree#1930] [Q&A] [answered]
```typescript
const RootStore = types.model({
  inbox: InboxStore,
  activeProject: types.maybeNull(
    types.reference(ProjectInbox, {
      get(identifier: string, parent /* HERE: What type do I use for `parent`? */) {
        return parent.inbox.projects.find(({ id }) => id == identifier) || null;
      },
      set(project) {
        return project.id;
      },
    })
  ),
});
```

If I set the type of `parent` to be `Instance<typeof RootStore>`, I get error: `'RootStore' implicitly has type 'any' because it does not have a type annotation and is referenced directly or indirectly in its own initializer.`.

PS: I am absolutely excited to have discovered this library.

**Comment by EmilTholin** [comment, marked as answer]
Hi @davidatsurge!

As you figured out TS sadly cannot figure out how to type `RootStore` if `RootStore` is used in its own definiton, so what you could do is type the `parent.inbox` to `Instance<typeof InboxStore>` instead to get around this.

**Example**
```ts
const RootStore = types.model({
  inbox: InboxStore,
  activeProject: types.maybeNull(
    types.reference(ProjectInbox, {
      get(identifier, parent) {
        if (parent === null) return null;

        const inbox = parent.inbox as Instance<typeof InboxStore>;

        return (
          inbox.projects.find(({ id }) => id === identifier) || (null as any)
        );
      },
      set(project) {
        return project.id;
      }
    })
  )
});
```

**Reply by davidatsurge** [reply]
Ahh, clever. Or at least I'm not that clever 😂. Thanks!
